### PR TITLE
Track per-request sql/http/log costs

### PR DESCRIPTION
**Pull request by bloodearnest**
This tracks the duration of each http/sql/log call in a per-request context, and adds that information to the access log.

It also refactors some record keeping, and tweaks a the gunicorn metrics a bit, as that was part of an earlier version of this branch

### CODE REVIEW

**Comment by bloodearnest** [comment]
This pull request **introduces 1 alert** when merging eeb45caf3e9bd1207af5b8ba44c3ce34b3617387 into 82eb65120ef7b499881689e158d7150191d2cb99 - [view on LGTM.com](https://lgtm.com/projects/g/canonical-ols/talisker/rev/pr-756b22461e096279df096ec4610f75f9918dac83)

**new alerts:**

* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: scripts/metrics.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        buckets = [(float(s[1]['le']), s[2]) for s in metric.samples if 'le' in s[1]]
+        buckets.sort()
+        deacc = []
+        for (b1, v1), (b2, v2) in reversed(list(zip(buckets, buckets[1:]))):
+            deacc.append(v2 - v1)
+        deacc.append(buckets[0][1])
+        line = sparklines(list(reversed(deacc)))[0]
+    elif metric.type == 'counter':
+        line = metric.samples[0][2]
+    output.append((metric.name, line))
```
Comment (sparkiegeek) [inline scripts/metrics.py]: what if there's a metric type which is neither histogram nor counter? I suggest:

```
else:
    raise ValueError("Unknown metric type {}".format(metric.type))
```

File: talisker/gunicorn.py [hunk 2]
```diff
@@ -176,9 +177,10 @@ def get_extra(self, resp, req, environ, request_time, status):
         extra['status'] = status
         if 'x-view-name' in headers:
             extra['view'] = headers['x-view-name']
-        extra['duration_ms'] = (
-            request_time.seconds * 1000 +
-            float(request_time.microseconds) / 1000
+        rt = request_time
+        extra['duration_ms'] = round(
+            rt.seconds * 1000 + float(rt.microseconds) / 1000,
```
Comment (sparkiegeek) [inline talisker/gunicorn.py]: `round(rt.total_seconds() * 1000, 3)` ?

**Comment by bloodearnest** [comment]
This pull request **introduces 1 alert** when merging eed7a6de4c647b88a64f1f1c5612d565c81331eb into 82eb65120ef7b499881689e158d7150191d2cb99 - [view on LGTM.com](https://lgtm.com/projects/g/canonical-ols/talisker/rev/pr-c7689ea88a830b60a65ead2883f4350f8c6a3940)

**new alerts:**

* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by bloodearnest** [comment]
This pull request **introduces 1 alert** when merging bd4b60240600ca9a60680c9acaceefb6336c3e0b into 82eb65120ef7b499881689e158d7150191d2cb99 - [view on LGTM.com](https://lgtm.com/projects/g/canonical-ols/talisker/rev/pr-18fd53ba6bfb770c08d884f4d16dd6437beeeac3)

**new alerts:**

* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: MANIFEST.in [hunk 3]
```diff
@@ -3,5 +3,6 @@ include README.rst AUTHORS.rst CONTRIBUTING.rst HISTORY.rst LICENSE
 recursive-include docs *.rst
 recursive-exclude docs *
 recursive-exclude tests *
+recursive-exclude scripts *
```
Comment (sparkiegeek) [inline MANIFEST.in]: left over from the sparklines change?

**Pull request merged**